apps/api/src/app/environments-v1/usecases/output-renderers/in-app-output-renderer.usecase.ts

```diff
@@ -1,87 +1,57 @@
+
     return {
       subject: inApp.subject,
       body: inApp.body,
       avatar: inApp.avatar,
-      primaryAction:
-        inApp.primaryAction &&
-        inApp.primaryAction.label &&
-        inApp.primaryAction.redirect &&
-        inApp.primaryAction.redirect.url
-          ? {
-              label: inApp.primaryAction.label,
-              redirect: {
-                url: inApp.primaryAction.redirect.url,
-                target: inApp.primaryAction.redirect.target as RedirectTargetEnum,
-              },
-            }
-          : undefined,
-      secondaryAction:
-        inApp.secondaryAction &&
-        inApp.secondaryAction.label &&
-        inApp.secondaryAction.redirect &&
-        inApp.secondaryAction.redirect.url
-          ? {
-              label: inApp.secondaryAction?.label,
-              redirect: {
-                url: inApp.secondaryAction?.redirect.url,
-                target: inApp.secondaryAction?.redirect.target as RedirectTargetEnum,
-              },
-            }
-          : undefined,
-      redirect:
-        inApp.redirect && inApp.redirect.url
-          ? {
-              url: inApp.redirect.url,
-              target: inApp.redirect.target as RedirectTargetEnum,
-            }
-          : undefined,
+      primaryAction: this.buildActionIfAllPartsAvailable(primaryAction),
+      secondaryAction: this.buildActionIfAllPartsAvailable(secondaryAction),
+      redirect: this.buildRedirect(redirect),
       data: inApp.data as Record<string, unknown>,
     };
   }
-}
-const RedirectTargetEnumSchema = z.enum(['_self', '_blank', '_parent', '_top', '_unfencedTop']);
 
-const InAppRenderOutputSchema = z.object({
-  subject: z.string().optional(),
-  body: z.string(),
-  avatar: z.string().optional(),
-  primaryAction: z
-    .object({
-      label: z.string().optional(),
-      redirect: z.object({
-        url: z.string().optional(),
-        target: RedirectTargetEnumSchema.optional(), // Optional target
-      }),
-    })
-    .optional(),
-  secondaryAction: z
-    .object({
-      label: z.string().optional(),
-      redirect: z.object({
-        url: z.string().optional(),
-        target: RedirectTargetEnumSchema.optional(), // Optional target
-      }),
-    })
-    .optional(), // Optional secondary action
-  data: z.record(z.unknown()).optional(), // Optional data
-  redirect: z
-    .object({
-      url: z.string().optional(),
-      target: RedirectTargetEnumSchema.optional(), // Optional target
-    })
-    .optional(),
-});
+  private buildRedirect(redirect?: InAppRedirectType) {
+    if (!(redirect && redirect.url)) {
+      return undefined;
+    }
+
+    return {
+      url: redirect.url,
+      target: redirect.target as RedirectTargetEnum,
+    };
+  }
+
+  private buildActionIfAllPartsAvailable(primaryAction?: InAppActionType) {
+    if (!(primaryAction && primaryAction.label && primaryAction.redirect && primaryAction.redirect.url)) {
+      return undefined;
```

Comment:
Ditto about null vs undefined

Reply:
same here, it's not the default, it's the render outcome